### PR TITLE
[SPARK-16046][DOCS] Aggregations in the Spark SQL programming guide

### PR DESCRIPTION
**Pull request by aokolnychyi**
## What changes were proposed in this pull request?

- A separate subsection for Aggregations under “Getting Started” in the Spark SQL programming guide. It mentions which aggregate functions are predefined and how users can create their own.
- Examples of using the `UserDefinedAggregateFunction` abstract class for untyped aggregations in Java and Scala.
- Examples of using the `Aggregator` abstract class for type-safe aggregations in Java and Scala.
- Python is not covered. 
- The PR might not resolve the ticket since I do not know what exactly was planned by the author. 

In total, there are four new standalone examples that can be executed via `spark-submit` or `run-example`. The updated Spark SQL programming guide references to these examples and does not contain hard-coded snippets. 

## How was this patch tested?

The patch was tested locally by building the docs. The examples were run as well. 

image

### CODE REVIEW

File: examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedUntypedAggregation.scala
Comment (assafmendelson) [inline examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedUntypedAggregation.scala]: Maybe add a little explanation here. For example, when I first saw this I tried to figure out where "salary" appears in the code as in practice it is being accessed by index only (input.getLong(0)).

File: examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedUntypedAggregation.scala
Comment (assafmendelson) [inline examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedUntypedAggregation.scala]: I believe an explanation on what MutableAggregationBuffer is should be added.
Basically explain how to access it, what it means for it to be mutable (including probably explaining that arrays and map types are immutable even if the buffer itself is mutable) etc.

File: examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedUntypedAggregation.scala
Comment (assafmendelson) [inline examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedUntypedAggregation.scala]: I would go with inputColumn. 
What I think should be more strongly explained is that this is basically the schema of the input for the aggregate function and not for the source dataframe.  Basically someone might think that their original dataframe might need to have this name for the column.

File: examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedTypedAggregation.java
Comment (marmbrus) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedTypedAggregation.java]: Its a little confusing to have the comment here for this optimization, but then not implement it.

File: examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedTypedAggregation.java
Comment (marmbrus) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedTypedAggregation.java]: I might be a little clearer if this was a `Person` with a `name` and `salary`.

File: examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedTypedAggregation.scala
Comment (marmbrus) [inline examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedTypedAggregation.scala]: Same comment here with object reuse.

File: examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedTypedAggregation.scala
Comment (marmbrus) [inline examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedTypedAggregation.scala]: Maybe comment what `name` is doing here.  I actually had to look it up.

**Comment by marmbrus** [comment]
This is great!  Thanks for taking the time to write up such complete examples.  I think this was a big gap in the existing docs.  One other ask.  The screen-shot is great, but I'd like to see which parts actually make it into the code snipits in the doc.  Ideally you could post a link to compiled doc.  If thats hard I can also try to build locally though.

**Comment by aokolnychyi** [comment]
@marmbrus I have updated the pull request. The compiled docs can be found [here](https://aokolnychyi.github.io/spark-docs/sql-programming-guide.html). 

I did not manage to build the Java API docs. I believe the problem is in my local installation. Therefore, I checked each url manually, they should work once the API docs are compiled. I will verify everything one more time in the nightly build.

File: docs/sql-programming-guide.md
Comment (jnh5y) [inline docs/sql-programming-guide.md]: As a suggestion, I'd change this to read:
"The [built-in DataFrames functions](api/scala/index.html#org.apache.spark.sql.functions$) provide common aggregations such as `count()`, `countDistinct()`, `avg()`, `max()`, and `min()`."

File: docs/sql-programming-guide.md
Comment (jnh5y) [inline docs/sql-programming-guide.md]: I think it'd be worth showing an Spark SQL example using the included/pre-defined functions.  Since your example implements 'avg', maybe use 'min' / 'max'?

Alternatively, the example could be added to the SQL statements in the main driver for the UserDefinedAggregateFunction implementations.

File: examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedUntypedAggregation.java
Comment (jnh5y) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedUntypedAggregation.java]: Should all UDAFs written in Java be static classes?  Similarly, should Scala implementations be Scala objects?

**Review by jnh5y — APPROVED** [review]
My comments are generally small suggestions; the PR looks great to me. 

I wonder if a discussion of the types and functions in a UDAF would be worthwhile.

Also, the code looks similar to the Hive test code here: https://github.com/apache/spark/blob/master/sql/hive/src/test/java/org/apache/spark/sql/hive/aggregate/MyDoubleAvg.java.

File: docs/sql-programming-guide.md
Comment (aokolnychyi) [inline docs/sql-programming-guide.md]: I also thought about this. In my view, it will be appropriate to have a separate subsection before Aggregations to show how to apply predefined SQL functions, including writing your own UDFs. That's will be worth another pull request. Alternatively, I can also try to extend this one to add an example of `max()` or `min()`. @marmbrus what's your opinion?

File: examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedUntypedAggregation.java
Comment (aokolnychyi) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedUntypedAggregation.java]: I tried to group all relevant entities for each example into one single class. Static classes are used here just to avoid the cumbersome code for creating the entities inside the `main()` method. Scala objects are also used due to a similar reason. I can mention this but it will add more comments to the already long examples.

**Comment by michalsenkyr** [comment]
If you are having trouble building Javadoc, try switching to Java 7 temporarily. Java 8 introduced stricter Javadoc rules that may fail the docs build. Unfortunately Jenkins doesn't, so new errors get introduced over time.

File: examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedTypedAggregation.scala
Comment (aokolnychyi) [inline examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedTypedAggregation.scala]: @michalsenkyr It is not required to create a new object in the `merge` method. One can modify the vars and return the existing object just like in the `reduce` method.  However, it is less critical here since this method will be called on pre-aggregated data and not for every element. On the one hand, I can apply here the same approach as in the `reduce` method to make the example consistent. On the other hand, the current code shows that it is not mandatory to modify vars. Probably, a comment might help. I am not sure which approach is better. Therefore, I am open to suggestions.

Comment (michalsenkyr) [inline examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedTypedAggregation.scala]: Personally, I prefer consistency. When I saw this, I immediately wondered whether there is a specific reason you did it this way.
I'd rather see both methods use the same paradigm. In this case probably the immutable one as the option of mutability is already mentioned in the comment above.
Or you can mention it again in the comment on this method if you want to provide examples of both. This way it just seems a little confusing.

File: examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedTypedAggregation.java
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedTypedAggregation.java]: Is this meant to be `MyAverage`?

**Review by srowen — COMMENTED** [review]
Except for one small question, the text/code looks OK. Others who know the content seem to approve, and you have gotten tests to pass, so seems reasonable.

File: examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedTypedAggregation.java
Comment (aokolnychyi) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedTypedAggregation.java]: @srowen `Average` is a Java bean that holds current sum and count. It is defined earlier. Here it represents a zero value. `MyAverage`, in turn, is the actual aggregator that accepts instances of the `Employee` class, stores intermediate results using an instance of`Average`, and produces `Double` as a result. 

I can rename `MyAverage` to `MyAverageAggregator` if this makes things clearer.

Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedTypedAggregation.java]: My bad, I read this incorrectly while skimming.

**Comment by marmbrus** [comment]
Sorry for the delay.  This LGTM, but I'm currently away from my Apache SSH keys.  Other committers should feel free to merge if you get there before I do.

**Comment by gatorsmile** [comment]
Sure, let me quickly go over the changes. Will merge it after that.

**Review by gatorsmile — APPROVED** [review]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.1

**Pull request closed without merging**
